### PR TITLE
Add optional test infrastructure

### PR DESCRIPTION
**Pull request by Xazax-hun**
This is a work in progress to get some feedback about the desired structure for tests.

Everything is subject to change including the testing framework used.

Some details:
- By default, the tests are not even compiled, no hard dependency is added.
- When BUILD_TESTING option is turned on, the code will check gtest out from GitHub and build it. This can be replaced with downloading a tarball if we prefer that. After turning on the testing, tests are built by default and there is a new target `make test` to run the tests.
- When gtest is found, no code will be downloaded or built. CMake will use the gtest that is installed on the system (using the distribution package manager, or conan, vcpkg, compiler from source, etc). A CMake option can be used to override this behavior and force downloading gtest.
- The CI jobs are modified to also run the tests. Note that the CI jobs first build the project without the tests and reconfigure the project to include the tests as an additional step. This way the CI also validates whether we can build the project without enabling the tests. It is also possible to consume this project without having gtest as a dependency. The user of IPR could use any test framework of its choice, there will be no conflicts there.
- The added test is very silly, I expect it to be replaced with more meaningful tests.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,12 +1,17 @@
 cmake_minimum_required(
-   VERSION 3.5
+   VERSION 3.9
```
Comment (Xazax-hun) [inline CMakeLists.txt:2]: Starting from version `3.9`, `CMake` added additional support for `gtest`, making the output of `make test` a bit nicer. 
Currently, `3.10` is the CMake version in Ubuntu 18.04, so this version would still support the two most recent LTS releases. It is not a must-have version bump, so it could be worked around if we wanted to.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
 )
 
 option(BUILD_DOC
    "Build LaTeX documentation"
    OFF
 )
 
+option(BUILD_TESTS
+   "Build the tests. Requires GTest."
+   OFF
```
Comment (MathiasMagnus) [inline CMakeLists.txt:12]: Strongly in favor of using canonical CMake variables. [`BUILD_TESTING`](https://cmake.org/cmake/help/latest/command/enable_testing.html) is the variable generally used for enabling CTests.

My guess is that IPR covers ground large enough to actually warrant a large battery of tests. Projects of similar size I've worked on used a combination of CTest wrapping Google Tests which are fairly large. Take [rocPRIM](https://github.com/ROCmSoftwarePlatform/rocPRIM/blob/develop/test/rocprim/CMakeLists.txt) for instance, where every `.cpp` test is actually a GTest with 10-100 tests inside. `block_scan` is one CTest that's instantiated with a handful of types, sizes, corner cases by GTest for a decent test matrix. It's a nice setup.

File: cmake/External_GTest.cmake [hunk 3]
```diff
@@ -0,0 +1,58 @@
+find_package(Threads REQUIRED)
+
+include(ExternalProject)
```
Comment (MathiasMagnus) [inline cmake/External_GTest.cmake:3]: Having been burned by scripts venturing into dependency management land, the question I've come to ask myself in such cases: who does the project want to cater to? Users who don't necessarily know how to obtain dependencies or users who are already using one (Vcpkg, Conan, etc.) and actually already have things built and a simple `find_package(GTest)` would simply succeed?

I recently engaged in an enlightening [dicsussion](https://discourse.cmake.org/t/depending-on-package-config-file-as-external-project/1501/10?u=mathiasmagnus) around this topic. I myself am deriving some collection of best practices, but CMake is the wild west and it's hard to select the best approach for any given project (requires extensive knowledge).

The AMD libraries still haven't decided on proper defaults and overrides your GTest installs without question. Every configuration time is aggravated by full builds of GTest and Google Bench, even if you have the best and latest installed. Once you start testing their availability, or adding "all or nothing" or specific self-hosting capabilities, complexity starts exploding, hence I'm usually against built-in dependency management when having a say. It's nice to cater to novice users, but don't screw advanced users over.

Comment (Xazax-hun) [reply cmake/External_GTest.cmake:3]: An alternative would be to try `find_package(GTest)` and download/build gtest only when it is not available. This might be the best of both worlds at the cost of some additional logic in the CMake file.

Comment (MathiasMagnus) [reply cmake/External_GTest.cmake:3]: Note, how rocPRIM is doing a similar approach (but it actually invokes the `install` target of deps inside [`Dependencies.cmake`](https://github.com/ROCmSoftwarePlatform/rocPRIM/blob/develop/cmake/Dependencies.cmake) which I'd gladly burn to the ground. It's code that shouldn't exist.

If going down this road, I'd minimally suggest:
- Something like `option(FORCE_DOWNLOAD_EXTERNAL)` which has some default.
- If the option is set, do the thing in the scripts currently
- If not, first try a `find_pacakge(GTest)` without `REQUIRED`
  - If present, merrily use it
  - If not, the existing machinery kicks in
- I'd invoke the deps `install` target at configure time, then I don't have to extract the essence of the Package Config file which you do now.
- My own `install` target would include copying over the deps installed dir from my build tree.

This is a fairly comprehensive, self-contained solution that is quite far from zero code, hence why I'd just point people in the direction of using Vcpkg and live a happy life. :)

**Comment by Xazax-hun** [comment]
I refined this PR based on some comments from @MathiasMagnus . Now it relies on a more standard CMAKE option, it builds GTest at the configuration stage and only builds it when the tests are on and it was not found on the user's machine.

**Comment by MathiasMagnus** [comment]
Looks good to me.

@GabrielDosReis I recently listened to you cppcast interview. The transition of MSVC to IPR sounds mighty interesting. When I encountered IPR, I got super excited over it's applications for reflection. (I'd gladly chat about it.) However, the docs being severely out of date I decided to target my limited "free" time elsewhere. Should the pracitce of unit testing take off, it's substitute parts of the docs where simple stuff could be picked up from. Newcomers could collect the bits and pieces they need to learn from the tests themselves.

File: cmake/External_GTest.cmake (hunk 3 above)
Comment (GabrielDosReis) [reply cmake/External_GTest.cmake:4]: @MathiasMagnus 
> who does the project want to cater to?

Any tool based on or consuming compiled module or header units should be able to use this; any compiler processing C++, on any platform with C++ compilers.

File: CMakeLists.txt (hunk 2 above)
Comment (GabrielDosReis) [reply CMakeLists.txt]: @MathiasMagnus 
> My guess is that IPR covers ground large enough to actually warrant a large battery of tests.

Agreed.  The issue isn't whether there should be tests; but sort of tests?  Just unit tests?  More integrated tests?  What is the appropriate framework?

Comment (MathiasMagnus) [reply CMakeLists.txt]: XPR<->IPR conversions are prime candidates for unit tests. As Xazax said, having examples and being part of CI would ensure that they never go out of date. Implementing a few typical transformations (const propagation, common sub-expression elimination, math transforms, etc.) would really help people get started.

Comment (GabrielDosReis) [reply CMakeLists.txt]: @MathiasMagnus - I hope we aren't talking past each other.  See the discussions on the test infrastructure for more context for my comment.

Comment (MathiasMagnus) [reply CMakeLists.txt]: @GabrielDosReis I hope so too. If by "appropriate framework" we're looking for some label we can attach to the test methodology, because IPR doesn't make much sense in isolation, "more integrated" tests without pulling in 3rd parties (not integration testing strictly speaking) pretty much translates to writing example code / documentation.

If "testing" were only as much as having all the exmples currently in the docs be available as standalone executables and checking that they actually compile (they don't), that'd still be better than nothing.

If by "appropriate framework" you meant how to actually design the test infrastructure, I'd unit test the hell out of IPR<->XPR conversion, as that truly is unit testable. Following that have as many usage examples as reasonably possible, all implementing some tranformation. Input and expected output dumped as XPR may be committed to the repo beside all of the exmples. What Xazax is proposing is to write the examples in a particular style, so they are as self-documenting and understandable as possible, mitigating the need for LaTeX docs which may go out-of-sync. I haven't seen codebases written that way, it may work.

I'm currently familiarizing myself with Doxygen+Sphinx hooked up to CMake and it's working well. It's more tech (maintenance), but has decent output. With Doxygen+Sphinx, it's trivial to embed the interesting parts of the examples (without boilerplate) into ReST docs and then compile to PDF, HTML, etc.

**Comment by GabrielDosReis** [comment]
Abstracting over the pretty printer issue, could you give me a quick summary of what the flow of checkout, build, tests, look like after this patch is in?

**Comment by Xazax-hun** [comment]
Sure!

This patch should be a no-op for those not interested in unit-tests. So checking out, building, consuming the library would be the same as before unless `-DBUILD_TESTING=ON` is passed to cmake at configuration time explicitly.

The process of checking out and running the tests is the following:
```
> git clone git@github.com:GabrielDosReis/ipr.git
> cd ipr
> mkdir build && cmake -DBUILD_TESTING=ON -Bbuild # This might take longer if GTest is not found on the host as it will be checked out and compiled.
> cd build
> make # This will now also build the tests.
> make test # This will run the tests.
```

I did not test it on Windows yet, but I do not anticipate any problems (let me know if there are any, and I will try to sort it out).

The test output looks like the following:
```
Running tests...
Test project .../ipr/build
    Start 1: TestIO.GlobalConstIntWithInit
1/1 Test #1: TestIO.GlobalConstIntWithInit ....***Exception: SegFault  0.38 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.43 sec

The following tests FAILED:
	  1 - TestIO.GlobalConstIntWithInit (SEGFAULT)
Errors while running CTest
make: *** [Makefile:149: test] Error 8
```

**Comment by GabrielDosReis** [comment]
Lovely.

> So checking out, building, consuming the library would be the same as before unless -DBUILD_TESTING=ON is passed to cmake at configuration time explicitly.

Could I persuade you to add this information into an `INSTALL` file?

**Comment by Xazax-hun** [comment]
Sure!

Let me know if you had something different in mind. Do you want me to remove/alter the failing test to keep the master green?

**Comment by GorNishanov** [comment]
Hi @Xazax-hun:

I am wondering if you have any experience with doctest: https://blog.jetbrains.com/rscpp/2019/07/10/better-ways-testing-with-doctest/ .

On the pro side: it has minimal footprint, just one header file.
On the cons side: no support for death tests, no built-in support for mocks

**Comment by Xazax-hun** [comment]
Hi Gor,

Do you envision including the header file in the repository?

Catch is another alternative that is header only.

The reason I started with GTest because that one has the widest support from IDEs, and it has good CMake integration. I am not very familiar with IPR yet, so I did not know what features are we interested in, starting with a feature rich solution sounded like a safe bet :)

Given the size of the project, I do not see the compilation time as an issue at this point. 

I do not have a strong preference for testing framework, and it should be very easy to migrate this PR to anything at this point. Do you want me to migrate to doctest?

**Comment by GorNishanov** [comment]
Yes. I was thinking of simply adding a doctest.h header to the repo. https://github.com/onqtam/doctest/blob/master/doc/markdown/build-systems.md

**Comment by Xazax-hun** [comment]
@GorNishanov I opened an alternative pull request here: https://github.com/GabrielDosReis/ipr/pull/69
I ended up adding the whole repo instead of the header only as I considered the CTest integration useful.
We could make it more slim by either dropping the CTest integration and only include the header file, or we can add the cmake files and the header separately without the rest of the repo (without the docs, examples and so on).

Alternatively, I can also convert the added repository to a submodule.

**Comment by Nostracodus** [comment]
@Xazax-hun I've never tried to integrate CTest and doctest. Can you explain why including the repo makes integration with CTest easier (for my own understanding)? My understanding of how I'd integrate with CTest was that I'd build an executable with all my tests, and then do something like `add_test(unittests <suitearguments>)` in the CMakeLists.txt

**Comment by GorNishanov** [comment]
I concur. doctest's benefit is being able to just add a single file to your repository. If we are adding 200+ files, we are losing the primary benefit of doctest (single header file unit testing framework).

**Comment by Xazax-hun** [comment]
@Nostracodus 

I wanted to use `doctest_discover_tests` which has some advantages. It discovers the tests at build time so we do not need to rerun cmake when new test cases are added. But you are right, we can just use add test and keep the repository leaner. 

https://github.com/onqtam/doctest/blob/master/scripts/cmake/doctest.cmake

**Comment by Xazax-hun** [comment]
To summarize the options to make the doctest PR leaner:
* Add header and cmake files instead of the full repo.
* Add header file only and do ctest integration manually. 
* Add header file only and do not care about ctest.

**Comment by GorNishanov** [comment]
Can you expand on the options 1 and 2?

For option 1, do we just need to add a single header file and a single cmake script?

For option 2, what is the burden of manual test maintenance?

**Comment by Xazax-hun** [comment]
Sure!

As far as I understand option 1 would involve adding one header file and two cmake files.

For option 2, it is about the granularity. 

With `add_test(unittests <suitearguments>)` we could either add this line once per binary, so CTest will only know about the binaries but do not know about the individual test cases in each binary. Or we could use the `suitearguments` so CTest know about each specific test case but it is a lot of work. The `doctest_discover_tests` version will parse the output of the `--list-test-cases` flag for each binary and build a list of individual test cases. After that CTest should be aware of the name of each individual test. As fas as I understand this can be used to run tests matching a regular expression and we can extract the test results from CTest with more granularity (for uploading to a dashboard if desired). 

Basically, we can do all of that by invoking the unit tests directly with the correct flags. But having an option to do most stuff from CTest might be handy for those who used to do things that way and it might be useful for building CIs with dashboards. Although this is definitely something I do NOT plan to look into :)

**Comment by Xazax-hun** [comment]
So one use case for the more granular understanding of tests for CTest would be:
```
ctest -R template
```

To run all tests that have anything to do with templates from all test binaries after a change that affects templates.

**Comment by GabrielDosReis** [comment]
I favor simplicity - small is beautiful, and IPR should be small.

I understand the rationale for going "full feature-rich from the get go" but I would rather be deliberate here and opt for minimalism.  I would like to just see one header file to be used for tests -- which I understood was a big selling point of doctest.

**Comment by Xazax-hun** [comment]
https://github.com/GabrielDosReis/ipr/pull/69 was updated to only include the single doctest header without the ctest integration.

**Comment by Xazax-hun** [comment]
#69 was merged, so I close this PR.
